### PR TITLE
middleware/proxy: absorb httpproxy

### PR DESCRIPTION
**Pull request by miekg**

### CODE REVIEW

File: middleware/proxy/README.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ proxy FROM TO... {
     health_check PATH:PORT [DURATION]
     except IGNORED_NAMES...
     spray
-    protocol [dns|https_google]
+    protocol [dns [ADDRESS...]|https_google]
```
Comment (johnbelamaric) [inline middleware/proxy/README.md]: What is ADDRESS here?

**Comment by miekg** [comment]
Will be a thing where you can specify the upstream for resolving
DNS.google.com or /etc/resolved.conf just like other address
specifications. Name could be better

On 15 Jan 2017 1:18 pm, "John Belamaric" <notifications@github.com> wrote:

*@johnbelamaric* commented on this pull request.
------------------------------

In middleware/proxy/README.md
<https://github.com/miekg/coredns/pull/481#pullrequestreview-16714158>:

> @@ -26,7 +26,7 @@ proxy FROM TO... {
     health_check PATH:PORT [DURATION]
     except IGNORED_NAMES...
     spray
-    protocol [dns|https_google]
+    protocol [dns [ADDRESS...]|https_google]

What is ADDRESS here?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/miekg/coredns/pull/481#pullrequestreview-16714158>, or mute
the thread

**Comment by johnbelamaric** [comment]
So, you would list protocol twice like this?

```
protocol dns /etc/resolv.conf
protocol https_google
```

That doesn't really make sense to me. What about something like:

```
protocol https_google
bootstrap dns [ADDRESS]
```

or

`protocol https_google bootstrap [ADDRESS]`

if it only applies to https_google (though I suppose it wouldn't).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] middleware/prox..." ]
>So, you would list protocol twice like this?
>
>```
>protocol dns /etc/resolv.conf
>protocol https_google
>```

Why can it be done twice?


>
>```
>protocol https_google
>bootstrap dns [ADDRESS]
>```
>
>or
>
>`protocol https_google bootstrap [ADDRESS]`
>
>if it only applies to https_google (though I suppose it wouldn't).

That's nice syntax. For now I am only adding `https_google`.

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
Exchanger needs probably to be moved to the staticUpstream and can then be shared between all HostPools

**Comment by miekg** [comment]
@johnbelamaric @chrisohaver Finally pushed some updates, could you guys give a (2nd) glance over - this could do with some review.
It still needs setup tests, fixing the conflicts and anything else I forgot.

Dropping 400 LOC is not bad though :-)

File: middleware/proxy/README.md [hunk 2]
```diff
@@ -112,3 +135,20 @@ proxy . /etc/resolv.conf {
...
+
+Proxy all requests within example.org to Google's dns.google.com.
+
+~~~
+proxy example.org dns.google.com {
+    protocol https_google
+}
+~~~
+
+Proxy everything, and re-lookup `dns.google.com` every 30 seconds using the resolvers specified
```
Comment (johnbelamaric) [inline middleware/proxy/README.md]: 30 or 300?

File: middleware/proxy/README.md [hunk 3]
```diff
@@ -53,18 +53,41 @@ available. This is to preeempt the case where the healthchecking (as a mechanism
...
+Currently `protocol` supports `dns` (i.e., standard DNS over UDP/TCP) and `https_google` (JSON
+payload over HTTPS). Note that with `https_google` the entire transport is encrypted. Only *you* and
+*Google* can see your DNS activity.
+
+* `dns`: no options can be given at the moment.
+* `https_google`: bootstrap **ADDRESS...** is used to (re-)resolve `dns.google.com` to an address to
+  connect to. This happens every 300s. If not specified the default is used: 8.8.8.8:53/8.8.4.4:53.
+
+  Debug queries are enabled by default and currently there is no way to turn them off. When CoreDNS
+  receives a debug queries (i.e. the name is prefixed with `o-o.debug.` a TXT record with Comment
```
Comment (johnbelamaric) [inline middleware/proxy/README.md]: s/queries/query/

Comment (johnbelamaric) [reply middleware/proxy/README.md]: Also missing closing paren after `o-o.debug.`.

File: middleware/proxy/README.md [hunk 4]
```diff
@@ -37,9 +37,9 @@ proxy FROM TO... {
 * `max_fails` is the number of failures within fail_timeout that are needed before considering a backend to be down. If 0, the backend will never be marked as down. Default is 1.
 * `health_check` will check path (on port) on each backend. If a backend returns a status code of 200-399, then that backend is healthy. If it doesn't, the backend is marked as unhealthy for duration and no requests are routed to it. If this option is not provided then health checks are disabled. The default duration is 10 seconds ("10s").
 * `ignored_names...` is a space-separated list of paths to exclude from proxying. Requests that match any of these paths will be passed through.
-* `spray` when all backends are unhealthy, randomly pick one to send the traffic to. (This is a failsafe.)
```
Comment (johnbelamaric) [inline middleware/proxy/README.md]: Are you removing this feature?

File: middleware/proxy/README.md (hunk 2 above)
Comment (johnbelamaric) [reply middleware/proxy/README.md]: README says 300

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
Looks good overall. Some nits here.

File: middleware/proxy/README.md (hunk 4 above)
Comment (miekg) [reply middleware/proxy/README.md]: No, $editor foobar. Fixed.

File: middleware/proxy/google.go [hunk 5]
```diff
@@ -178,6 +194,49 @@ func extractAnswer(m *dns.Msg) ([]string, error) {
...
+		from:        from,
+		Hosts:       nil,
+		Policy:      &Random{},
+		Spray:       nil,
+		FailTimeout: 10 * time.Second,
+		MaxFails:    3,
+		ex:          ex,
+	}
+
+	fmt.Printf("new hosts %v\n", hosts)
```
Comment (johnbelamaric) [inline middleware/proxy/google.go]: remove debug

**Comment by miekg** [comment]
Thanks. All done. I'll add some more tests.

**Review by johnbelamaric — APPROVED** [review]

**Comment by miekg** [comment]
I'll merge as-is and fixup the tests in later PRs

**Pull request merged**